util/priority_queue.go

```diff
@@ -100,7 +134,176 @@ func (pq *PriorityQueue) Dequeue() Op {
 		return nil
 	}
 
-	op := heap.Pop(&pq.queue).(Op)
-	delete(pq.hit, op.Key())
-	return op
+	item := heap.Pop(&pq.queue).(*item)
+	delete(pq.hit, item.payload.Key())
+	return item.payload
+}
+
+// DelayedCall is a function that we're not going to run yet.
+type DelayedCall struct {
+	clock      clockwork.Clock
+	f          func()
+	elapsed    <-chan time.Time
+	cancelled  chan struct{}
+	terminated chan struct{}
+}
+
+// DelayCall runs 'f' after 'delay'.
+func DelayCall(clock clockwork.Clock, delay time.Duration, f func()) *DelayedCall {
+	call := DelayedCall{
+		clock: clock,
+		f:     f,
+	}
+	call.reset(delay)
+	return &call
+}
+
+func (d *DelayedCall) loop() {
+	defer close(d.terminated)
+	for {
+		select {
+		case <-d.elapsed:
+			d.f()
+		case <-d.cancelled:
+			return
+		}
+	}
+}
+
+// Cancel the delayed call.
+func (d *DelayedCall) Cancel() {
+	close(d.cancelled)
+	<-d.terminated
+}
+
+func (d *DelayedCall) reset(delay time.Duration) {
+	d.cancelled = make(chan struct{})
+	d.terminated = make(chan struct{})
+	d.elapsed = d.clock.After(delay)
+	go d.loop()
+}
+
+// Reset the delayed call to run after 'delay' (as measured from now).
+func (d *DelayedCall) Reset(delay time.Duration) {
+	d.Cancel()
+	d.reset(delay)
+}
+
+// ScheduledItem is an item in a queue of scheduled items.
+type ScheduledItem interface {
+	Key() string
+	// Scheduled returns the earliest possible time the time is available for
+	// dequeueing.
+	Scheduled() time.Time
+}
+
+// scheduledOp adapts a ScheduledItem to an Op
+type scheduledOp struct {
+	ScheduledItem
+}
+
+// Priority implements Op.
+func (op scheduledOp) Priority() int64 {
+	return -op.Scheduled().Unix()
+}
+
+// SchedulingQueue is like a priority queue, but the first item is the oldest
+// scheduled item.
+type SchedulingQueue struct {
```

Comment:
Ta. I think local metrics are better (what if you want to have more than one scheduled queue?) but have gone for global to reduce review friction and also because it's less work.

Adding support for de-duping meant a chunk of repeated logic around adding an item to the queue. At first I defined an `addItem` function inside the go routine, but then in 79e5338 I extracted that to a method on a new internal struct that tracks queue state.